maro/simulator/scenarios/data_center/business_engine.py

```diff
@@ -0,0 +1,320 @@
+        # Update all live VMs CPU utilization.
+        self._update_vm_util()
+        # Update all PM CPU utilization.
+        self._update_pm_util()
+        # TODO
+        # Generate VM requirement events from data file.
+        # It might be implemented by a for loop to process VMs in each tick.
+        vm_requirement_event = self._event_buffer.gen_cascade_event(
+            tick=tick,
+            event_type=Events.REQUIREMENTS,
+            payload=None
+        )
+        self._event_buffer.insert_event(event=vm_requirement_event)
+
+    def get_metrics(self) -> DocableDict:
+        """Get current environment metrics information.
+
+        Returns:
+            DocableDict: Metrics information.
+        """
+
+        return DocableDict(
+            metrics_desc,
+            energy_consumption=self._energy_consumption,
+            success_requirements=self._successful_requirements,
+            failed_requirements=self._failed_requirements,
+            total_latency=self._total_latency
+        )
+
+    def _register_events(self):
+        # Register our own events and their callback handlers.
+        self._event_buffer.register_event_handler(event_type=Events.REQUIREMENTS, handler=self._on_vm_required)
+        self._event_buffer.register_event_handler(event_type=Events.FINISHED, handler=self._on_vm_finished)
+
+        # Generate decision event.
+        self._event_buffer.register_event_handler(event_type=MaroEvents.TAKE_ACTION, handler=self._on_action_received)
+
+    def _update_vm_util(self):
+        """Update all live VMs CPU utilization.
+
+        The length of VMs utilization series could be difference among all VMs,
+        because index 0 represents the VM's CPU utilization at the tick it starts.
+        """
+
+        for vm in self._live_vm.values():
+            vm.util_cpu = vm.get_util(cur_tick=self._tick)
+
+    def _update_pm_util(self):
+        """Update CPU utilization occupied by total VMs on each PM."""
+
+        for pm in self._machines:
+            total_util_cpu: int = 0
+            for vm_id in pm.vm_set:
+                vm = self._live_vm[vm_id]
+                total_util_cpu += vm.util_cpu * vm.req_cpu / 100
+            util_cpu = total_util_cpu / pm.cap_cpu * 100
+            pm.update_util(tick=self._tick, util_cpu=util_cpu)
+            self._energy_consumption += self._cpu_util_to_energy_consumption(pm.util_cpu)
+
+    def _cpu_util_to_energy_consumption(self, cpu_util: float) -> float:
+        """Convert the CPU utilization to energy consumption."""
+
+        power: float = self._conf["calibration parameter"]
+        # NOTE: Energy comsumption parameters should refer to more research.
+        busy_power = self._conf["busy_power"]
+        idle_power = self._conf["idle_power"]
+        return idle_power + (busy_power - idle_power) * (2 * cpu_util - pow(cpu_util, power))
+
+    def _postpone_vm_requirement(self, postpone_type: PostponeType, vm_id: int, remaining_buffer_time: int):
```

Comment:
rename suggestion for buffer time:

initial_time_budget
remaining_time_budget